Fix retains grad behavior after in-place

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81114
* #81055
* __->__ #79996

See this doc: https://docs.google.com/document/d/1KiRdnoj6B4cI3yl017hTbCqcOGO1gWIpUf20sldipHM/edit#

Two issues (1) regarding hooks in general and (2) regarding retains grad hooks are fixed, Python hooks, which rely on a different mechanism are not discussed here:
- Hooks in cpp in general
  - (fixed) new hooks to registered to a newer version of the tensor no longer get applied to grad_fn
    associated with older version of the tensor when the first hook was ever registered
  - (unchanged) hooks registered to the older version of the tensor remain active on
- Retains grad hooks
  - (fixed) now get moved to the latest grad_fn. NB: To the user, retains_grad is not considered hooks
    or expected to behave like hooks (which we consider properties of the grad_fn) vs retains_gradness
    which is a property of the tensor.
- (not in this PR) Python hooks
  - (will fix) same issue as hooks in cpp where new hooks are being applied to grad_fn associated
    with the older version of the tensor

soulitzer committed:
Fix retains grad behavior after in-place

[ghstack-poisoned]

soulitzer commented on test/cpp/api/autograd.cpp:
```diff
@@ -831,6 +831,71 @@ TEST(CustomAutogradTest, Hooks) {
   ASSERT_THROWS_WITH(y.remove_hook(3), "Invalid index");
 }
 
+TEST(CustomAutogradTest, HooksInplace) {
```
Test that behavior of cpp hooks are not affected

soulitzer commented on test/test_autograd.py:
```diff
@@ -836,6 +836,18 @@ def test_retain_grad(self):
         out.backward()
         self.assertEqual(input * 18, input.grad)
 
+    def test_retain_grad_inplace_over_view(self):
```
TODO: add more tests for python

albanD commented on torch/csrc/autograd/variable.h:
```diff
@@ -222,8 +222,10 @@ struct TORCH_API AutogradMeta : public c10::AutogradMetaInterface {
-  // Only meaningful on non-leaf variables (must be false otherwise)
-  bool retains_grad_;
+  // Only meaningful on non-leaf variables (must be -1 otherwise)
+  // The value of retains_grad_ indicates the index of it in cpp_hooks_list_
```
nit: This is more a handle than an index right?

soulitzer replied:
What is the difference?

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -181,6 +200,8 @@ void rebase_history(const Variable& self, Edge gradient_edge) {
 
   set_gradient_edge(self, std::move(gradient_edge));
+  // Pass both self and its grad_fn to avoid calling into grad_fn reentrantly
+  torch::autograd::impl::update_retains_grad(self, self.grad_fn());
```
Shouldn't think be done for the view case as well?

soulitzer replied:
We discussed this offline already, but just to document it here, we want to have it in the grad_fn instead of rebase history because other views that weren't directly updated in-place will also be affected

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -154,6 +154,25 @@ AutogradMeta* materialize_autograd_meta(const at::TensorBase& self) {
+    return;
+  }
+  auto idx = impl::get_autograd_meta(self)->retains_grad_;
+  auto old_list = materialize_autograd_meta(self)->cpp_hooks_list_;
```
Not need to materialize here. You segfaulted on the line above if it wasn't there already.

soulitzer replied:
Good catch, somehow missed this lol

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -154,6 +154,25 @@ AutogradMeta* materialize_autograd_meta(const at::TensorBase& self) {
+  (*old_list)[idx] = nullptr;
+  materialize_autograd_meta(self)->cpp_hooks_list_ = new_list;
+
+  std::unique_ptr<FunctionPreHook> hook_ptr(
```
Why not reset the `AM.retains_grad_` field and call `self.retain_grad()` here? That will avoid duplication.

soulitzer replied:
retain_grad would call into register_hook which would wrap the function in a lambda again

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -154,6 +154,25 @@ AutogradMeta* materialize_autograd_meta(const at::TensorBase& self) {
+  }
+  auto idx = impl::get_autograd_meta(self)->retains_grad_;
+  auto old_list = materialize_autograd_meta(self)->cpp_hooks_list_;
+  auto retains_grad_hook = (*old_list)[idx];
```
Index in a list is bad! What happens if someone removed elements from it??

soulitzer committed:
Update on "Fix retains grad behavior after in-place"


See this doc: https://docs.google.com/document/d/1KiRdnoj6B4cI3yl017hTbCqcOGO1gWIpUf20sldipHM/edit#

Two issues (1) regarding hooks in general and (2) regarding retains grad hooks are fixed, Python hooks, which rely on a different mechanism are not discussed here:
- Hooks in cpp in general
  - (fixed) new hooks to registered to a newer version of the tensor no longer get applied to grad_fn
    associated with older version of the tensor when the first hook was ever registered
  - (unchanged) hooks registered to the older version of the tensor remain active on
- Retains grad hooks
  - (fixed) now get moved to the latest grad_fn. NB: To the user, retains_grad is not considered hooks
    or expected to behave like hooks (which we consider properties of the grad_fn) vs retains_gradness
    which is a property of the tensor.
- (not in this PR) Python hooks
  - (will fix) same issue as hooks in cpp where new hooks are being applied to grad_fn associated
    with the older version of the tensor

[ghstack-poisoned]

soulitzer committed:
Update on "Fix retains grad behavior after in-place"


See this doc: https://docs.google.com/document/d/1KiRdnoj6B4cI3yl017hTbCqcOGO1gWIpUf20sldipHM/edit#

Two issues (1) regarding hooks in general and (2) regarding retains grad hooks are fixed, Python hooks, which rely on a different mechanism are not discussed here:
- Hooks in cpp in general
  - (fixed) new hooks to registered to a newer version of the tensor no longer get applied to grad_fn
    associated with older version of the tensor when the first hook was ever registered
  - (unchanged) hooks registered to the older version of the tensor remain active on
- Retains grad hooks
  - (fixed) now get moved to the latest grad_fn. NB: To the user, retains_grad is not considered hooks
    or expected to behave like hooks (which we consider properties of the grad_fn) vs retains_gradness
    which is a property of the tensor.
- (not in this PR) Python hooks
  - (will fix) same issue as hooks in cpp where new hooks are being applied to grad_fn associated
    with the older version of the tensor

[ghstack-poisoned]

soulitzer committed:
Update on "Fix retains grad behavior after in-place"


See this doc: https://docs.google.com/document/d/1KiRdnoj6B4cI3yl017hTbCqcOGO1gWIpUf20sldipHM/edit#

Two issues (1) regarding hooks in general and (2) regarding retains grad hooks are fixed, Python hooks, which rely on a different mechanism are not discussed here:
- Hooks in cpp in general
  - (fixed) new hooks to registered to a newer version of the tensor no longer get applied to grad_fn
    associated with older version of the tensor when the first hook was ever registered
  - (unchanged) hooks registered to the older version of the tensor remain active on
- Retains grad hooks
  - (fixed) now get moved to the latest grad_fn. NB: To the user, retains_grad is not considered hooks
    or expected to behave like hooks (which we consider properties of the grad_fn) vs retains_gradness
    which is a property of the tensor.
- (not in this PR) Python hooks
  - (will fix) same issue as hooks in cpp where new hooks are being applied to grad_fn associated
    with the older version of the tensor

[ghstack-poisoned]

soulitzer committed:
Update on "Fix retains grad behavior after in-place"


See this doc: https://docs.google.com/document/d/1KiRdnoj6B4cI3yl017hTbCqcOGO1gWIpUf20sldipHM/edit#

Two issues (1) regarding hooks in general and (2) regarding retains grad hooks are fixed, Python hooks, which rely on a different mechanism are not discussed here:
- Hooks in cpp in general
  - (fixed) new hooks to registered to a newer version of the tensor no longer get applied to grad_fn
    associated with older version of the tensor when the first hook was ever registered
  - (unchanged) hooks registered to the older version of the tensor remain active on
- Retains grad hooks
  - (fixed) now get moved to the latest grad_fn. NB: To the user, retains_grad is not considered hooks
    or expected to behave like hooks (which we consider properties of the grad_fn) vs retains_gradness
    which is a property of the tensor.
- (not in this PR) Python hooks
  - (will fix) same issue as hooks in cpp where new hooks are being applied to grad_fn associated
    with the older version of the tensor

[ghstack-poisoned]

soulitzer committed:
Update on "Fix retains grad behavior after in-place"


See this doc: https://docs.google.com/document/d/1KiRdnoj6B4cI3yl017hTbCqcOGO1gWIpUf20sldipHM/edit#

Two issues (1) regarding hooks in general and (2) regarding retains grad hooks are fixed, Python hooks, which rely on a different mechanism are not discussed here:
- Hooks in cpp in general
  - (fixed) new hooks to registered to a newer version of the tensor no longer get applied to grad_fn
    associated with older version of the tensor when the first hook was ever registered
  - (unchanged) hooks registered to the older version of the tensor remain active on
- Retains grad hooks
  - (fixed) now get moved to the latest grad_fn. NB: To the user, retains_grad is not considered hooks
    or expected to behave like hooks (which we consider properties of the grad_fn) vs retains_gradness
    which is a property of the tensor.
- (not in this PR) Python hooks
  - (will fix) same issue as hooks in cpp where new hooks are being applied to grad_fn associated
    with the older version of the tensor

[ghstack-poisoned]

soulitzer committed:
Update on "Fix retains grad behavior after in-place"


See this doc: https://docs.google.com/document/d/1KiRdnoj6B4cI3yl017hTbCqcOGO1gWIpUf20sldipHM/edit#

Two issues (1) regarding hooks in general and (2) regarding retains grad hooks are fixed, Python hooks, which rely on a different mechanism are not discussed here:
- Hooks in cpp in general
  - (fixed) new hooks to registered to a newer version of the tensor no longer get applied to grad_fn
    associated with older version of the tensor when the first hook was ever registered
  - (unchanged) hooks registered to the older version of the tensor remain active on
- Retains grad hooks
  - (fixed) now get moved to the latest grad_fn. NB: To the user, retains_grad is not considered hooks
    or expected to behave like hooks (which we consider properties of the grad_fn) vs retains_gradness
    which is a property of the tensor.
- (not in this PR) Python hooks
  - (will fix) same issue as hooks in cpp where new hooks are being applied to grad_fn associated
    with the older version of the tensor

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -154,6 +154,25 @@ AutogradMeta* materialize_autograd_meta(const at::TensorBase& self) {
   return get_autograd_meta(self);
 }
 
+void update_retains_grad(
```
This is actually updating both hooks and retains_grad right?

soulitzer replied:
Renamed to "update_cpp_hooks_on_new_gradfn"

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -154,6 +154,25 @@ AutogradMeta* materialize_autograd_meta(const at::TensorBase& self) {
+  new_list->push_back(std::move((*meta->cpp_hooks_list_)[idx]));
+  (*meta->cpp_hooks_list_)[idx] = nullptr;
+  meta->cpp_hooks_list_ = new_list;
+  meta->retains_grad_ = 0;
```
nit: small comment here that since this is a new list, 0 is the index of the retains_grad hook

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -154,6 +154,25 @@ AutogradMeta* materialize_autograd_meta(const at::TensorBase& self) {
   return get_autograd_meta(self);
 }
 
+void update_retains_grad(
```
There are a couple assumption for this function that are worth mentioning I think:
- The AM is already populated
- The fn does exist (not AccumulateGrad here)
- The fn has no hooks already set

albanD reviewed (approved):
SGTM
Small nits only!

soulitzer commented on test/cpp/api/autograd.cpp:
```diff
@@ -831,6 +831,111 @@ TEST(CustomAutogradTest, Hooks) {
+    ASSERT_VARIABLE_EQ(grad, torch::ones({5, 5}) * 2);
+  });
+
+  bool hook2_count = 0;
```
whoops!

soulitzer committed:
Update on "Fix retains grad behavior after in-place"


See this doc: https://docs.google.com/document/d/1KiRdnoj6B4cI3yl017hTbCqcOGO1gWIpUf20sldipHM/edit#

Two issues (1) regarding hooks in general and (2) regarding retains grad hooks are fixed, Python hooks, which rely on a different mechanism are not discussed here:
- Hooks in cpp in general
  - (fixed) new hooks to registered to a newer version of the tensor no longer get applied to grad_fn
    associated with older version of the tensor when the first hook was ever registered
  - (unchanged) hooks registered to the older version of the tensor remain active on
- Retains grad hooks
  - (fixed) now get moved to the latest grad_fn. NB: To the user, retains_grad is not considered hooks
    or expected to behave like hooks (which we consider properties of the grad_fn) vs retains_gradness
    which is a property of the tensor.
- (not in this PR) Python hooks
  - (will fix) same issue as hooks in cpp where new hooks are being applied to grad_fn associated
    with the older version of the tensor

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g
